### ISSUE TITLE
Define a deprecation policy for Falcon's API

### ISSUE THREAD

**Issue by sebasmagri** [falconry/falcon#542]
We should review how other projects are handling the deprecation of API methods and document one for us.
## To be defined
- [ ] Grace period for deprecated items: how much releases will still have the deprecated item
- [ ] Requirements to deprecate an API item: an item should not be deprecated if there is not a clear migration path for developers.
- [ ] Mechanism for notification: raise warnings during the grace period? Raise exceptions afterwards? Document deprecated items in release notes or is a separated deprecation schedule needed?.
## Policies to Review
- [ ] Django: https://docs.djangoproject.com/en/1.8/internals/release-process
- [ ] Rallydev: https://help.rallydev.com/web-services-api-deprecation-policy
- [X] Raven: http://raven.readthedocs.org/en/latest/#deprecation-notes
- [X] PBR's SemVer Deprecation Guidelines: http://docs.openstack.org/developer/pbr/semver.html#how-should-i-handle-deprecating-functionality

**Comment by weldpua2008** [comment]
raise warnings during the grace period

 Raise exceptions afterwards

 Document deprecated items in release notes or is a separated deprecation schedule needed

**Comment by richardolsson** [comment]
I think [semantic versioning](http://semver.org/) would be a good idea. This project is still young but is starting to gain traction and use, so it would probably be a good idea to jump to 1.0.0 soonish, and then use the major version number to indicate backwards incompatible changes, and use minor releases to deprecate features.

Semantic versioning has it's problems, but for a library used only by developers (as opposed to end users) a consistent and "standardized" (oh well) versioning scheme could go a long way in complementing a proper deprecation scheme.

Note that according to semantic version anything is allowed to happen before 1.0.0. In practice however, falcon is already being used in production even though it's just 0.3.0, so it might make sense to be a little careful but make the jump sooner rather than later.

**Comment by tbug** [comment]
I'm with @richardolsson on having deprecation duration be integrated with version numbering.
semver or not, binding deprecation to version numbers allow an easy way to freeze if you are not yet ready to update.

I'm thinking something like this would be a nice way to go about it:

Before any major release, mark feature as deprecated in docstring (and write at what version the feature will stop working), write in changelog AND a dedicated deprecation schedule, AND do a minor release with the deprecation.
Apart from the docstring and documentation, nothing has changed yet.

Then for the following major releases (major, as in semver):

1. Start emitting warnings (via logging, prererably muteable) about use of deprecated function/feature, but otherwise still functioning. Emitted message should contain some hint as to how to move away from the deprecated feature.
2. Raise an error with directions on how to migrate. Feature no longer functioning.
3. Remove completely. Devs will get `NameError` and the likes.


So, if a feature is deprecated in major version 1, it will continue as always for the duration of 1, emit warnings in major version 2, and raise errors in 3. In 4 it will be completely removed.

Personally I like semver, but it is religion. Any well-defined versioning is fine.

Imo. Django's versioning is horrible for anything but time-defined releases, and it doesn't seem like falcon will ever have a pre-defined release cycle (?).


tl;dr: Imo: deprecations should keep functioning for the duration of the major release they were made in, emit warnings in the next. Then break.

**Comment by oscarcp** [comment]
+1 @richardolsson & @tbug I think semver is the way to go.

**Comment by AxisOfEval** [comment]
+1 @richardolsson and @tbug All for semver.

**Comment by vytas7** [comment]
Closing this as our documentation does lay out a policy that is clear enough for the end users IMHO:
> **Reliable**. We go to great lengths to avoid introducing breaking changes, and when we do they are fully documented and only introduced (in the spirit of [SemVer](http://semver.org/)) with a major version increment. The code is rigorously tested with numerous inputs and we require 100% coverage at all times. Falcon does not depend on any external Python packages.

Informally, we do also follow what @tbug has suggested above, although exceptions were sometimes made for small breaking changes before a major release, like when having to accommodate ASGI in 3.0. I'll add a note on https://github.com/falconry/falcon/issues/473.
